File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.js
const fs = require('fs')
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n')

const [n, m] = input[0].split(' ').map(Number) // 물품의 수 n, 버틸 수 있는 무게 m
let dp = Array.from({ length: m + 1 }, () => 0)

for (let i = 1; i <= n; i++){
  const [w, v] = input[i].split(' ').map(Number) // 물품의 무게 w, 가치 v
  for (let j = m; j >= w; j--){
    dp[j] = Math.max(dp[j], dp[j-w]+v)
  }
}

console.log(dp[m])

File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.py
import sys 
input = sys.stdin.readline 

n, k = map(int, input().split())  # 필요한 n개의 물건, 최대 무게 k
dp = [0] * (k + 1)

for _ in range(n):
    w, v = map(int, input().split())  # 물건의 무게 w, 가치 v
    for i in range(k, w - 1, -1):  # 가능한 무게 역순으로 순회
        dp[i] = max(dp[i], dp[i - w] + v)

print(dp[k])
